hedera-node/src/main/java/com/hedera/services/ServicesState.java

```diff
@@ -363,6 +371,30 @@ public RecordsRunningHashLeaf runningHashLeaf() {
 		return metadata.getUniqueTreasuryOwnershipAssociations();
 	}
 
+	private void updateLinks() {
+		var accounts = accounts();
+		var tokenRels = tokenAssociations();
+
+		for (var accountId : accounts.keySet()) {
+			var merkleAccount = accounts.getForModify(accountId);
+
+			EntityNumPair prevListRootKey = MISSING_NUM_PAIR;
+			for (var tokenId : merkleAccount.tokens().asTokenIds()) {
+				var newListRootKey = EntityNumPair.fromLongs(accountId.longValue(), tokenId.getTokenNum());
+				var association = tokenRels.getForModify(newListRootKey);
+				association.setNextKey(prevListRootKey);
+
+				if (prevListRootKey != MISSING_NUM_PAIR) {
+					var prevAssociation = tokenRels.getForModify(prevListRootKey);
+					prevAssociation.setPrevKey(newListRootKey);
+				}
+				prevListRootKey = newListRootKey;
+			}
+			merkleAccount.setLastAssociatedToken(prevListRootKey);
```

Comment:
Here we should do `merkleAccount.setAssociatedTokensCount(merkleAccount.tokens().numAssociations())`, right?